File: docker/extensions/word-count.ts
import { Type } from "@sinclair/typebox";

export default function (pi: any) {
  pi.registerTool({
    name: "word_count",
    label: "Word Count",
    description: "Fast file analysis: returns line count, word count, character count, and file size. Use this instead of reading entire files when you just need basic stats.",
    parameters: Type.Object({
      path: Type.String({ description: "Path to the file to analyze" }),
    }),
    async execute(_toolCallId: string, params: { path: string }, signal: AbortSignal) {
      try {
        const result = await pi.exec(
          "python3",
          ["/usr/local/bin/bark-tools/word_count.py", params.path],
          { signal, timeout: 10000 }
        );
        return {
          content: [{ type: "text", text: result.stdout || result.stderr }],
          details: {},
        };
      } catch (err: any) {
        return {
          content: [{ type: "text", text: `Error: ${err.message}` }],
          details: {},
        };
      }
    },
  });
}

File: docker/tools/word_count.py
#!/usr/bin/env python3
"""Analyze a text file: line count, word count, character count, file size."""
import sys
import os

def analyze(path):
    if not os.path.exists(path):
        print(f"Error: {path} not found")
        sys.exit(1)

    size = os.path.getsize(path)
    with open(path, 'r', errors='replace') as f:
        content = f.read()

    lines = content.count('\n') + (1 if content and not content.endswith('\n') else 0)
    words = len(content.split())
    chars = len(content)

    print(f"File: {path}")
    print(f"Size: {size} bytes")
    print(f"Lines: {lines}")
    print(f"Words: {words}")
    print(f"Characters: {chars}")

if __name__ == '__main__':
    if len(sys.argv) != 2:
        print("Usage: word_count <file>")
        sys.exit(1)
    analyze(sys.argv[1])
